File: packages/ui-compose/ui/src/main/kotlin/com/kinetixui/ui/Textarea.kt
package com.kinetixui.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsFocusedAsState
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * KinetixTextarea — mirrors `packages/ui/src/components/textarea.tsx`,
 * which is (per its own doc comment) identical to [KinetixInput] except
 * multi-line with `min-h-[100px]`. Same token choices as `KinetixInput`;
 * 100dp is the Figma-literal min-height, not on the shared `spacing_*`
 * scale — hardcoded, same reasoning as `KinetixBadge`'s padding.
 */
@Composable
fun KinetixTextarea(
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    placeholder: String? = null,
    enabled: Boolean = true,
    isError: Boolean = false,
) {
    val colors = KinetixColorScheme.current
    val interactionSource = remember { MutableInteractionSource() }
    val isFocused by interactionSource.collectIsFocusedAsState()

    val borderColor = when {
        isError -> colors.destructive
        isFocused -> colors.primary
        else -> colors.border
    }
    val textStyle = TextStyle(
        color = if (enabled) colors.foreground else colors.mutedForeground,
        fontSize = dimensionResource(R.dimen.font_size_body_md).value.sp,
        lineHeight = dimensionResource(R.dimen.line_height_body_md).value.sp,
        letterSpacing = dimensionResource(R.dimen.letter_spacing_wide_25).value.sp,
    )
    val shape = RoundedCornerShape(dimensionResource(R.dimen.radius_sm))

    Box(
        modifier = modifier
            .heightIn(min = 100.dp)
            .background(colors.background, shape)
            .border(dimensionResource(R.dimen.border_width_default), borderColor, shape)
            .padding(dimensionResource(R.dimen.spacing_3)),
    ) {
        if (value.isEmpty() && placeholder != null) {
            Text(text = placeholder, style = textStyle.copy(color = colors.mutedForeground))
        }
        BasicTextField(
            value = value,
            onValueChange = onValueChange,
            modifier = Modifier,
            enabled = enabled,
            singleLine = false,
            textStyle = textStyle,
            interactionSource = interactionSource,
            cursorBrush = SolidColor(colors.primary),
        )
    }
}

File: packages/ui-swiftui/Sources/KinetixUI/Textarea.swift
//
// Textarea.swift — KinetixTextarea.
//
// Mirrors packages/ui/src/components/textarea.tsx — per its own doc
// comment, identical to KinetixInput except multi-line with
// `min-h-[100px]`. Same token choices; 100 is the Figma-literal
// min-height, off the shared spacing scale.
//
// Built on `TextEditor`, which has no native placeholder — one is
// overlaid when empty. `TextEditor`'s internal text inset isn't publicly
// specified, so the placeholder alignment is approximate.
//

import SwiftUI

public struct KinetixTextarea: View {
    @Environment(\.kinetixColors) private var colors
    @Environment(\.isEnabled) private var isEnabled
    @FocusState private var focused: Bool

    @Binding private var text: String
    private let placeholder: String
    private let isError: Bool

    public init(text: Binding<String>, placeholder: String = "", isError: Bool = false) {
        self._text = text
        self.placeholder = placeholder
        self.isError = isError
    }

    private var borderColor: Color {
        if isError { return colors.destructive }
        return focused ? colors.primary : colors.input
    }

    public var body: some View {
        TextEditor(text: $text)
            .focused($focused)
            .font(.system(size: 14))
            .foregroundStyle(colors.foreground)
            .scrollContentBackground(.hidden) // iOS 16 / macOS 13
            .frame(minHeight: 100, alignment: .topLeading)
            .padding(12) // spacing/3
            .background(colors.background, in: RoundedRectangle(cornerRadius: 4, style: .continuous))
            .overlay(alignment: .topLeading) {
                if text.isEmpty {
                    Text(placeholder)
                        .font(.system(size: 14))
                        .foregroundStyle(colors.mutedForeground)
                        .padding(.top, 20)
                        .padding(.leading, 17)
                        .allowsHitTesting(false)
                }
            }
            .overlay {
                RoundedRectangle(cornerRadius: 4, style: .continuous)
                    .strokeBorder(borderColor, lineWidth: 1)
            }
            .opacity(isEnabled ? 1 : 0.5)
            .animation(.easeInOut(duration: 0.12), value: focused)
    }
}
